README.md
# guige-
test


harfbuzz/ChangeLog
commit 3e86feb54c94e46b60168fd2a4773183eb1354b4
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Sat Jan 21 19:40:30 2012 -0500

    Speed up colorless diff

 test/shaping/hb_test_tools.py |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

commit 1e58df603457cb4b57da78b5d4a8df66aa7d7be4
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Sat Jan 21 19:37:31 2012 -0500

    Cleanup manifest code

 test/shaping/hb_test_tools.py |   18 ++++++++++++------
 1 files changed, 12 insertions(+), 6 deletions(-)

commit 956d552e108eeb50bb3ad21588830af7a2f3862a
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Sat Jan 21 19:31:51 2012 -0500

    Port hb-manifest-update to Python

 test/shaping/hb-manifest-update |   23 +++--------------------
 test/shaping/hb_test_tools.py   |   38
 ++++++++++++++++++++++++++++++--------
 2 files changed, 33 insertions(+), 28 deletions(-)

commit 3a34e9e351ed0ee3eb27f9c0f154bc227f1226bf
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Sat Jan 21 19:15:41 2012 -0500

    Ignore Broken Pipe errors

 test/shaping/hb_test_tools.py |   76
 ++++++++++++++++++++++++++---------------
 1 files changed, 48 insertions(+), 28 deletions(-)

commit d5300241680844f5625f32792f7dd7181ed05f9b
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Sat Jan 21 19:07:22 2012 -0500

    [util] Make clusters work with char offset instead of UTF-8 offset

    This means the --features indices also refer to char position
    instead of byte position now.  Same for cluster values reported
    by hb-shape.

    Will add an option for byte indices later.

 util/options.cc |    7 +++++--
 util/options.hh |   12 ++++++++++++
 2 files changed, 17 insertions(+), 2 deletions(-)

commit 0f68f4a0b5ee78cbdb2a89a9a1a9125afe72ed2f
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Sat Jan 21 18:59:48 2012 -0500

    Correctly print out Unicode strings

 util/options.cc |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit f22089ac24f43ff048c2a0f1c1f604ae3a96be8b
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Fri Jan 20 21:21:47 2012 -0500

    Misc fixes

 test/shaping/hb-diff          |    8 ++++----
 test/shaping/hb_test_tools.py |    4 ++--
 2 files changed, 6 insertions(+), 6 deletions(-)

commit 96968bfae5ce61a0a098bf0e6acd2210a309a499
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Fri Jan 20 21:16:34 2012 -0500

    Port hb-manifest-read to Python

 test/shaping/hb-diff-filter-failures |    2 +-
 test/shaping/hb-manifest-read        |   37 ++-----------------------
 test/shaping/hb_test_tools.py        |   49
 ++++++++++++++++++++++++++++++---
 3 files changed, 48 insertions(+), 40 deletions(-)

commit a59ed46fa4f7b76605f8ce6e75783ead406468f5
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Fri Jan 20 20:56:32 2012 -0500

    Add final residues from test-shape-complex

 .../in-tree/shaper-arabic/script-nko/misc/MANIFEST |    1 +
 .../in-tree/shaper-arabic/script-nko/misc/misc.txt |    5 +++++
 test/shaping/texts/in-tree/shaper-default/MANIFEST |    1 +
 .../shaper-default/script-linear-b/MANIFEST        |    1 +
 .../shaper-default/script-linear-b/misc/MANIFEST   |    1 +
 .../shaper-default/script-linear-b/misc/misc.txt   |    1 +
 test/shaping/texts/in-tree/shaper-indic/MANIFEST   |    1 +
 .../in-tree/shaper-indic/south-east-asian/MANIFEST |    1 +
 .../south-east-asian/script-khmer/MANIFEST         |    1 +
 .../south-east-asian/script-khmer/misc/MANIFEST    |    1 +
 .../south-east-asian/script-khmer/misc/misc.txt    |    9 +++++++++
 11 files changed, 23 insertions(+), 0 deletions(-)

commit 820e0ed318d9b187a131baa9491d5d390ec33ef4
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Fri Jan 20 20:51:52 2012 -0500

    Add Punjabi tests from test-shape-complex also

 .../indic/script-punjabi/misc/MANIFEST             |    1 +
 .../indic/script-punjabi/misc/misc.txt             |    1 +
 2 files changed, 2 insertions(+), 0 deletions(-)

commit a7d71c105772fb612871b4cae59bdae47bbc8751
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Fri Jan 20 20:50:09 2012 -0500

    Add Tamil test data from Muguntharaj Subramanian

 .../shaper-indic/indic/script-tamil/misc/MANIFEST  |    1 +
 .../shaper-indic/indic/script-tamil/misc/misc.txt  |   43
 ++++++++++++++++++++
 2 files changed, 44 insertions(+), 0 deletions(-)

commit 5992a9941e7f19181df1e34e79e514ccedc3d284
Author: Behdad Esfahbod <behdad@behdad.org>
Date:   Fri Jan 20 20:48:14 2012 -0500

    Import test data from late test-shape-complex

 .../in-tree/shaper-arabic/script-arabic/MANIFEST   |    1 +
 .../in-tree/shaper-arabic/script-mandaic/MANIFEST  |    1 +
 .../shaper-arabic/script-mongolian/MANIFEST        |    1 +
 .../in-tree/shaper-arabic/script-nko/MANIFEST      |    1 +
 .../in-tree/shaper-arabic/script-syriac/MANIFEST   |    1 +
 .../shaper-indic/indic/script-assamese/MANIFEST    |    1 +
 .../shaper-indic/indic/script-bengali/MANIFEST     |    1 +
 .../indic/script-bengali/misc/MANIFEST             |    1 +
 .../indic/script-bengali/misc/misc.txt             |   48
 ++++++++++++++++++++
 .../shaper-indic/indic/script-devanagari/MANIFEST  |    1 +
 .../indic/script-devanagari/misc/MANIFEST          |    1 +
 .../indic/script-devanagari/misc/misc.txt          |   23 +++++++++
 .../shaper-indic/indic/script-gujarati/MANIFEST    |    1 +
 .../shaper-indic/indic/script-kannada/MANIFEST     |    1 +
 .../indic/script-kannada/misc/MANIFEST             |    1 +
 .../indic/script-kannada/misc/misc.txt             |   17 +++++++
 .../shaper-indic/indic/script-malayalam/MANIFEST   |    1 +
 .../indic/script-malayalam/misc/MANIFEST           |    1 +
 .../indic/script-malayalam/misc/misc.txt           |   22 +++++++++
 .../shaper-indic/indic/script-oriya/MANIFEST       |    1 +
 .../shaper-indic/indic/script-oriya/misc/MANIFEST  |    1 +
 .../shaper-indic/indic/script-oriya/misc/misc.txt  |   28 +++++++++++
 .../shaper-indic/indic/script-punjabi/MANIFEST     |    1 +

ucdn/ucdn.h
/*
 * Copyright (C) 2012 Grigori Goronzy <greg@kinoho.net>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#ifndef UCDN_H
#define UCDN_H

#ifdef _MSC_VER
// Some versions of Dev Studio don't have stdint.h...
typedef unsigned __int32 uint32_t;
#else
#include <stdint.h>
#endif

#define UCDN_EAST_ASIAN_F 0
#define UCDN_EAST_ASIAN_H 1
#define UCDN_EAST_ASIAN_W 2
#define UCDN_EAST_ASIAN_NA 3
#define UCDN_EAST_ASIAN_A 4
#define UCDN_EAST_ASIAN_N 5

#define UCDN_SCRIPT_COMMON 0
#define UCDN_SCRIPT_LATIN 1
#define UCDN_SCRIPT_GREEK 2
#define UCDN_SCRIPT_CYRILLIC 3
#define UCDN_SCRIPT_ARMENIAN 4
#define UCDN_SCRIPT_HEBREW 5
#define UCDN_SCRIPT_ARABIC 6
#define UCDN_SCRIPT_SYRIAC 7
#define UCDN_SCRIPT_THAANA 8
#define UCDN_SCRIPT_DEVANAGARI 9
#define UCDN_SCRIPT_BENGALI 10
#define UCDN_SCRIPT_GURMUKHI 11
#define UCDN_SCRIPT_GUJARATI 12
#define UCDN_SCRIPT_ORIYA 13
#define UCDN_SCRIPT_TAMIL 14
#define UCDN_SCRIPT_TELUGU 15
#define UCDN_SCRIPT_KANNADA 16
#define UCDN_SCRIPT_MALAYALAM 17
#define UCDN_SCRIPT_SINHALA 18
#define UCDN_SCRIPT_THAI 19
#define UCDN_SCRIPT_LAO 20
#define UCDN_SCRIPT_TIBETAN 21
#define UCDN_SCRIPT_MYANMAR 22
#define UCDN_SCRIPT_GEORGIAN 23
#define UCDN_SCRIPT_HANGUL 24
#define UCDN_SCRIPT_ETHIOPIC 25
#define UCDN_SCRIPT_CHEROKEE 26
#define UCDN_SCRIPT_CANADIAN_ABORIGINAL 27
#define UCDN_SCRIPT_OGHAM 28
#define UCDN_SCRIPT_RUNIC 29
#define UCDN_SCRIPT_KHMER 30
#define UCDN_SCRIPT_MONGOLIAN 31
#define UCDN_SCRIPT_HIRAGANA 32
#define UCDN_SCRIPT_KATAKANA 33
#define UCDN_SCRIPT_BOPOMOFO 34
#define UCDN_SCRIPT_HAN 35
#define UCDN_SCRIPT_YI 36
#define UCDN_SCRIPT_OLD_ITALIC 37
#define UCDN_SCRIPT_GOTHIC 38
#define UCDN_SCRIPT_DESERET 39
#define UCDN_SCRIPT_INHERITED 40
#define UCDN_SCRIPT_TAGALOG 41
#define UCDN_SCRIPT_HANUNOO 42
#define UCDN_SCRIPT_BUHID 43
#define UCDN_SCRIPT_TAGBANWA 44
#define UCDN_SCRIPT_LIMBU 45
#define UCDN_SCRIPT_TAI_LE 46
#define UCDN_SCRIPT_LINEAR_B 47
#define UCDN_SCRIPT_UGARITIC 48
#define UCDN_SCRIPT_SHAVIAN 49
#define UCDN_SCRIPT_OSMANYA 50
#define UCDN_SCRIPT_CYPRIOT 51
#define UCDN_SCRIPT_BRAILLE 52
#define UCDN_SCRIPT_BUGINESE 53
#define UCDN_SCRIPT_COPTIC 54
#define UCDN_SCRIPT_NEW_TAI_LUE 55
#define UCDN_SCRIPT_GLAGOLITIC 56
#define UCDN_SCRIPT_TIFINAGH 57
#define UCDN_SCRIPT_SYLOTI_NAGRI 58
#define UCDN_SCRIPT_OLD_PERSIAN 59
#define UCDN_SCRIPT_KHAROSHTHI 60
#define UCDN_SCRIPT_BALINESE 61
#define UCDN_SCRIPT_CUNEIFORM 62
#define UCDN_SCRIPT_PHOENICIAN 63
#define UCDN_SCRIPT_PHAGS_PA 64
#define UCDN_SCRIPT_NKO 65
#define UCDN_SCRIPT_SUNDANESE 66
#define UCDN_SCRIPT_LEPCHA 67
#define UCDN_SCRIPT_OL_CHIKI 68
#define UCDN_SCRIPT_VAI 69
#define UCDN_SCRIPT_SAURASHTRA 70
#define UCDN_SCRIPT_KAYAH_LI 71
#define UCDN_SCRIPT_REJANG 72
#define UCDN_SCRIPT_LYCIAN 73
#define UCDN_SCRIPT_CARIAN 74
#define UCDN_SCRIPT_LYDIAN 75
#define UCDN_SCRIPT_CHAM 76
#define UCDN_SCRIPT_TAI_THAM 77
#define UCDN_SCRIPT_TAI_VIET 78
#define UCDN_SCRIPT_AVESTAN 79
#define UCDN_SCRIPT_EGYPTIAN_HIEROGLYPHS 80
#define UCDN_SCRIPT_SAMARITAN 81
#define UCDN_SCRIPT_LISU 82
#define UCDN_SCRIPT_BAMUM 83
#define UCDN_SCRIPT_JAVANESE 84
#define UCDN_SCRIPT_MEETEI_MAYEK 85
#define UCDN_SCRIPT_IMPERIAL_ARAMAIC 86
#define UCDN_SCRIPT_OLD_SOUTH_ARABIAN 87
#define UCDN_SCRIPT_INSCRIPTIONAL_PARTHIAN 88
#define UCDN_SCRIPT_INSCRIPTIONAL_PAHLAVI 89
#define UCDN_SCRIPT_OLD_TURKIC 90
#define UCDN_SCRIPT_KAITHI 91
#define UCDN_SCRIPT_BATAK 92
#define UCDN_SCRIPT_BRAHMI 93
#define UCDN_SCRIPT_MANDAIC 94
#define UCDN_SCRIPT_CHAKMA 95
#define UCDN_SCRIPT_MEROITIC_CURSIVE 96
#define UCDN_SCRIPT_MEROITIC_HIEROGLYPHS 97
#define UCDN_SCRIPT_MIAO 98
#define UCDN_SCRIPT_SHARADA 99
#define UCDN_SCRIPT_SORA_SOMPENG 100
#define UCDN_SCRIPT_TAKRI 101
#define UCDN_SCRIPT_UNKNOWN 102

#define UCDN_GENERAL_CATEGORY_CC 0
#define UCDN_GENERAL_CATEGORY_CF 1
#define UCDN_GENERAL_CATEGORY_CN 2
#define UCDN_GENERAL_CATEGORY_CO 3
#define UCDN_GENERAL_CATEGORY_CS 4
#define UCDN_GENERAL_CATEGORY_LL 5
#define UCDN_GENERAL_CATEGORY_LM 6
#define UCDN_GENERAL_CATEGORY_LO 7
#define UCDN_GENERAL_CATEGORY_LT 8
#define UCDN_GENERAL_CATEGORY_LU 9
#define UCDN_GENERAL_CATEGORY_MC 10
#define UCDN_GENERAL_CATEGORY_ME 11
#define UCDN_GENERAL_CATEGORY_MN 12
#define UCDN_GENERAL_CATEGORY_ND 13
#define UCDN_GENERAL_CATEGORY_NL 14
#define UCDN_GENERAL_CATEGORY_NO 15
#define UCDN_GENERAL_CATEGORY_PC 16
#define UCDN_GENERAL_CATEGORY_PD 17
#define UCDN_GENERAL_CATEGORY_PE 18
#define UCDN_GENERAL_CATEGORY_PF 19
#define UCDN_GENERAL_CATEGORY_PI 20
#define UCDN_GENERAL_CATEGORY_PO 21
#define UCDN_GENERAL_CATEGORY_PS 22
#define UCDN_GENERAL_CATEGORY_SC 23
#define UCDN_GENERAL_CATEGORY_SK 24
#define UCDN_GENERAL_CATEGORY_SM 25
#define UCDN_GENERAL_CATEGORY_SO 26
#define UCDN_GENERAL_CATEGORY_ZL 27
#define UCDN_GENERAL_CATEGORY_ZP 28
#define UCDN_GENERAL_CATEGORY_ZS 29

#define UCDN_BIDI_CLASS_L 0
#define UCDN_BIDI_CLASS_LRE 1
#define UCDN_BIDI_CLASS_LRO 2
#define UCDN_BIDI_CLASS_R 3
#define UCDN_BIDI_CLASS_AL 4
#define UCDN_BIDI_CLASS_RLE 5
#define UCDN_BIDI_CLASS_RLO 6
#define UCDN_BIDI_CLASS_PDF 7
#define UCDN_BIDI_CLASS_EN 8
#define UCDN_BIDI_CLASS_ES 9
#define UCDN_BIDI_CLASS_ET 10
#define UCDN_BIDI_CLASS_AN 11
#define UCDN_BIDI_CLASS_CS 12
#define UCDN_BIDI_CLASS_NSM 13
#define UCDN_BIDI_CLASS_BN 14
#define UCDN_BIDI_CLASS_B 15
#define UCDN_BIDI_CLASS_S 16
#define UCDN_BIDI_CLASS_WS 17
#define UCDN_BIDI_CLASS_ON 18

/**
 * Return version of the Unicode database.
 *
 * @return Unicode database version
 */
const char *ucdn_get_unicode_version(void);

/**
 * Get combining class of a codepoint.
 *
 * @param code Unicode codepoint
 * @return combining class value, as defined in UAX#44
 */
int ucdn_get_combining_class(uint32_t code);

/**
 * Get east-asian width of a codepoint.
 *
 * @param code Unicode codepoint
 * @return value according to UCDN_EAST_ASIAN_* and as defined in UAX#11.
 */
int ucdn_get_east_asian_width(uint32_t code);

/**
 * Get general category of a codepoint.
 *
 * @param code Unicode codepoint
 * @return value according to UCDN_GENERAL_CATEGORY_* and as defined in
 * UAX#44.
 */
int ucdn_get_general_category(uint32_t code);

/**
 * Get bidirectional class of a codepoint.
 *
 * @param code Unicode codepoint
 * @return value according to UCDN_BIDI_CLASS_* and as defined in UAX#44.
 */
int ucdn_get_bidi_class(uint32_t code);

/**
 * Get script of a codepoint.
 *
 * @param code Unicode codepoint
 * @return value according to UCDN_SCRIPT_* and as defined in UAX#24.
 */
int ucdn_get_script(uint32_t code);

/**
 * Check if codepoint can be mirrored.
 *
 * @param code Unicode codepoint
 * @return 1 if mirrored character exists, otherwise 0
 */
int ucdn_get_mirrored(uint32_t code);

/**
 * Mirror a codepoint.
 *
 * @param code Unicode codepoint
 * @return mirrored codepoint or the original codepoint if no
 * mirrored character exists
 */
uint32_t ucdn_mirror(uint32_t code);

/**
 * Pairwise canonical decomposition of a codepoint. This includes
 * Hangul Jamo decomposition.
 *
 * @param code Unicode codepoint
 * @param a filled with first codepoint of decomposition
 * @param b filled with second codepoint of decomposition, or 0
 * @return success
 */
int ucdn_decompose(uint32_t code, uint32_t *a, uint32_t *b);

/**
 * Compatibility decomposition of a codepoint.
 *
 * @param code Unicode codepoint
 * @param decomposed filled with decomposition, must be able to hold 18
 * characters
 * @return length of decomposition or 0 in case none exists
 */
int ucdn_compat_decompose(uint32_t code, uint32_t *decomposed);

/**
 * Pairwise canonical composition of two codepoints. This includes
 * Hangul Jamo composition.
 *
 * @param code filled with composition
 * @param a first codepoint
 * @param b second codepoint
 * @return success
 */
int ucdn_compose(uint32_t *code, uint32_t a, uint32_t b);

#endif
